src/core/Akka.Streams/Implementation/Fusing/Ops.cs

```diff
@@ -3822,24 +3815,55 @@ public override void OnPull()
                         // if result is false, it means enumerator was closed. Complete stage in that case.
                         CompleteStage();
                     }
+                } 
+                else if (vtask.IsCompleted) // IsCompleted covers Faulted, Cancelled, and RanToCompletion async state
+                {
+                    // vtask will always contains an exception because we know we're not successful and always throws
+                    try
+                    {
+                        // This does not block because we know that the task already completed
+                        // Using GetAwaiter().GetResult() to automatically unwraps AggregateException inner exception
+                        vtask.GetAwaiter().GetResult();
+                    }
+                    catch (Exception ex)
+                    {
+                        FailStage(ex);
+                        return;
+                    }
+
+                    throw new InvalidOperationException("Should never reach this code");
                 }
                 else
                 {
-                    vtask.AsTask().ContinueWith(_handleContinuation);
+                    async Task ProcessTask()
+                    {
+                        // Since this Action is used as task continuation, we cannot safely call corresponding
+                        // OnSuccess/OnFailure/OnComplete methods directly. We need to do that via async callbacks.
+                        try
+                        {
+                            var completed = await vtask.ConfigureAwait(false);
+                            if (completed)
+                                _onSuccess(_enumerator.Current);
+                            else
+                                _onComplete();
+                        }
+                        catch (Exception ex)
+                        {
+                            _onFailure(ex);
+                        }
+                    }
+
+#pragma warning disable CS4014
+                    ProcessTask();
+#pragma warning restore CS4014
                 }
             }
+
             public override void OnDownstreamFinish(Exception cause)
             {
-                var vtask = _enumerator.DisposeAsync();
-                if (vtask.IsCompletedSuccessfully)
-                {
-                    CompleteStage(); // if dispose completed immediately, complete stage directly
-                }
-                else
-                {
-                    // for async disposals use async callback
-                    vtask.GetAwaiter().OnCompleted(_onComplete);
-                }
+                _completionCts.Cancel();
+                _completionCts.Dispose();
```

Comment:
Do we have guarantees that cancelling will dispose the enumerator if needed?

Reply:
Oh, you're totally correct, I forgot that we need to dispose the enumerator... I guess the previous code is still needed here.

Reply:
```
            public override void OnDownstreamFinish(Exception cause)
            {
                _completionCts.Cancel();
                _completionCts.Dispose();
                CompleteStage();
                _enumerator.DisposeAsync().AsTask().ContinueWith(t => _baseOnDownstreamFinishCallback(cause));
            }
```
or
```
            public override void OnDownstreamFinish(Exception cause)
            {
                _completionCts.Cancel();
                _completionCts.Dispose();
                CompleteStage();

                _enumerator.DisposeAsync().GetAwaiter().GetResult();
                base.OnDownstreamFinish(cause);
            }
```

Reply:
I will add the try...catch later, just wanted to get the point across first

Reply:
I think we should go with the latter - I doubt there will be `IAsyncEnumerator.DisposeAsync` implementations that can block for prolonged periods of time.